### ISSUE TITLE
BUG: integrate: `solve_ivp` and `odeint` with `lsoda` have conflicting and unclear requirements for banded Jacobians

### ISSUE THREAD

**Issue by MatthewFlamm** [scipy/scipy#10793]
The lsoda method in solve_ivp does not honor uband and lband for banded jacobians.

#### Reproducing code example:
lsoda will never check the jacobian for a nonstiff problem, so I used the classical stiff Robertson ode and added trivial equations around it to make it banded.

odeint has no problem with the banded jacobian, but there is an error using solve_ivp.

```
from scipy.integrate import solve_ivp, odeint
import numpy as np

def f(t, y):
    return [
        y[0],
        -0.04 * y[1] + 1e4 * y[2] * y[3],
        0.04 * y[1] - 1e4 * y[2] * y[3] - 3e7 * y[2]**2,
        3e7 * y[2]**2,
        y[4]
    ]

def jac(t, y):
    return [
        [1,     0,                    0,         0, 0],
        [0, -0.04,             1e4*y[3],  1e4*y[2], 0],
        [0,  0.04, -1e4*y[3]-3e7*2*y[2], -1e4*y[2], 0],
        [0,     0,           3e7*2*y[2],         0, 0],
        [0,     0,                    0,         0, 1]
    ]

def band_jac(t, y):
    return [
        [0,     0,                    0,  1e4*y[2], 0],
        [0,     0,             1e4*y[3], -1e4*y[2], 0],
        [1, -0.04, -1e4*y[3]-3e7*2*y[2],         0, 1],
        [0,  0.04,           3e7*2*y[2],         0, 0]
    ]

rtol = 1e-8
atol = 1e-10

tspan = (0, 1000)
t_eval = np.linspace(*tspan, 100)
y0 = np.array([0, 1, 0, 0, 0])
method = 'LSODA'

sol = solve_ivp(f, tspan, y0, t_eval=t_eval, method='LSODA',
 rtol=rtol, atol=atol)
sol_jac = solve_ivp(f, tspan, y0, jac=jac, t_eval=t_eval, me
thod='LSODA', rtol=rtol, atol=atol)
assert np.allclose(sol.y, sol_jac.y)

y_odeint = odeint(f, y0, t_eval, tfirst=True, Dfun=jac)
assert np.allclose(sol.y, y_odeint.T)

y_odeint_band = odeint(f, y0, t_eval, tfirst=True, Dfun=band
_jac, ml=1, mu=2)
assert np.allclose(sol.y, y_odeint_band.T)

sol = solve_ivp(f, tspan, y0, lband=1, uband=2, jac=band_jac
, t_eval=t_eval, method='LSODA', rtol=rtol, atol=atol)
# error
```

#### Error message:

In this example it is looking for a matrix with 5 rows instead of the banded packed form matrix with 4 rows.   Note that this is the size of the full dense jacobian.  In my code (not given) where I was trying to debug the behavior, I have a much larger matrix and it also is looking for the full dense jacobian rather than the smaller packed one.

```
File "path/venv/lib/py
thon3.7/site-packages/scipy/integrate/_ode.py", line 1343, i
n run
    y1, t, istate = self.runner(*args)
ValueError: 0-th dimension must be fixed to 5 but got 4
```

#### Scipy/Numpy/Python version information:
```
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)

1.3.1 1.17.1 sys.version_info(major=3, minor=7, micro=3, rel
easelevel='final', serial=0)
```

I haven't been able to pinpoint the source of this yet.

**Comment by MatthewFlamm** [comment]
I have a hunch that these parameters should be converted to `ml` and `mu`  at some point in the function chain, but aren't.  If this is true, should we just change the parameter names to `ml` and `mu`?  This wouldn't really be a breaking change given that no one can be using the current names.

**Comment by MatthewFlamm** [comment]
This was wishful thinking.  Solve_ivp wraps the ode implementation, so I need to add this to the comparison.

**Comment by MatthewFlamm** [comment]
Updated code:. Neither solve_ivp nor ode works as advertised in the documentation.


```
from scipy.integrate import solve_ivp, odeint, ode
import numpy as np

def f(t, y):
    return [
        y[0],
        -0.04 * y[1] + 1e4 * y[2] * y[3],
        0.04 * y[1] - 1e4 * y[2] * y[3] - 3e7 * y[2]**2,
        3e7 * y[2]**2,
        y[4]
    ]

def jac(t, y):
    return [
        [1,     0,                    0,         0, 0],
        [0, -0.04,             1e4*y[3],  1e4*y[2], 0],
        [0,  0.04, -1e4*y[3]-3e7*2*y[2], -1e4*y[2], 0],
        [0,     0,           3e7*2*y[2],         0, 0],
        [0,     0,                    0,         0, 1]
    ]

def band_jac(t, y):
    return [
        [0,     0,                    0,  1e4*y[2], 0],
        [0,     0,             1e4*y[3], -1e4*y[2], 0],
        [1, -0.04, -1e4*y[3]-3e7*2*y[2],         0, 1],
        [0,  0.04,           3e7*2*y[2],         0, 0]
    ]

rtol = 1e-8
atol = 1e-10

tspan = (0, 1000)
t_eval = np.linspace(*tspan, 100)
y0 = np.array([0, 1, 0, 0, 0])
method = 'LSODA'

sol = solve_ivp(f, tspan, y0, t_eval=t_eval, method='LSODA',
 rtol=rtol, atol=atol)
sol_jac = solve_ivp(f, tspan, y0, jac=jac, t_eval=t_eval, me
thod='LSODA', rtol=rtol, atol=atol)
assert np.allclose(sol.y, sol_jac.y)

y_odeint = odeint(f, y0, t_eval, tfirst=True, Dfun=jac)
assert np.allclose(sol.y, y_odeint.T)

y_odeint_band = odeint(f, y0, t_eval, tfirst=True, Dfun=band
_jac, ml=1, mu=2)
assert np.allclose(sol.y, y_odeint_band.T)

r = ode(f, jac).set_integrator('lsoda')
r.set_initial_value(y0)
sol_ode=[]
for t in t_eval[1:]:
    sol_ode.append(r.integrate(t))
    assert r.successful()
assert np.allclose(sol.y[:, 1:], np.array(sol_ode).T)

r = ode(f, band_jac).set_integrator('lsoda', lband=1, uband=
2)
r.set_initial_value(y0)
sol_ode_band = []
for t in t_eval[1:]:
    sol_ode_band.append(r.integrate(t))
    # error
    assert r.successful()
assert np.allclose(sol.y[:, 1:], np.array(sol_ode_band).T)


sol = solve_ivp(f, tspan, y0, lband=1, uband=2, jac=band_jac
, t_eval=t_eval, method='LSODA', rtol=rtol, atol=atol)
# error
```

**Comment by MatthewFlamm** [comment]
I thought maybe that the transpose of the jacobian was wanted in ode lsoda, so I tried that too (code at bottom).  It now gives a different error:

```
ValueError: 1-th dimension must be fixed to 5 but got 4
```

To me, this clearly says that it wants the full, dense matrix and is not using the uband and lband parameters to ask for the packed, banded matrix.

```
def trans_band_jac(t,y):
    return np.array([
        [0,     0,                    0,  1e4*y[2], 0],
        [0,     0,             1e4*y[3], -1e4*y[2], 0],
        [1, -0.04, -1e4*y[3]-3e7*2*y[2],         0, 1],
        [0,  0.04,           3e7*2*y[2],         0, 0]
    ]).T

r = ode(f, trans_band_jac).set_integrator('lsoda', lband=1, uband=2)
r.set_initial_value(y0)
sol_ode_band = []
for t in t_eval[1:]:
    sol_ode_band.append(r.integrate(t))
    # error
    assert r.successful()
assert np.allclose(sol.y[:, 1:], np.array(sol_ode_band).T)
```

**Comment by MatthewFlamm** [comment]
This has been in the code for 6+ years? I might be using it wrong and it is a documentation thing. 

But, I'm noticing that the test for banded solvers for ode integrator uses a nonstiff example, which wouldn't actually test the banded nature of the stiff solver in LSODA.  That is, LSODA will never call the jacobian function for nonstiff problems, so an error is never raised.


Nonstiff jacobian in banded test, located [here](https://github.com/scipy/scipy/blob/master/scipy/integrate/tests/test_banded_ode_solvers.py#L138-L143):

```
# lband = 2, uband = 1:
    a_real = np.array([[-0.6, 0.1, 0.0, 0.0, 0.0],
                       [0.2, -0.5, 0.9, 0.0, 0.0],
                       [0.1, 0.1, -0.4, 0.1, 0.0],
                       [0.0, 0.3, -0.1, -0.9, -0.3],
                       [0.0, 0.0, 0.1, 0.1, -0.7]])
```

**Comment by MatthewFlamm** [comment]
Is this related to the fixes that were needed here? https://github.com/scipy/scipy/pull/3607

**Comment by kmaitreys** [comment]
Did you manage to find any workaround?

**Comment by MatthewFlamm** [comment]
The workaround for me was to use `odeint` instead of `solve_ivp` or the `ode` implementation.

**Comment by kmaitreys** [comment]
Ah thanks for the reply. I tried that, but unfortunately for my case, ```odeint``` is very slow and does not converge (I cannot understand why though).

**Comment by MatthewFlamm** [comment]
It could also be that I did not do a close enough performance comparison to see if it truly was working in odeint.  It did integrate successfully at least.

**Comment by ivan-pi** [comment]
Just stumbled upon this error. The solution by @MatthewFlamm to add an extra empty row seems to work. It is consistent at least with the workspace requirements of the LU factorization routine [`DGBTRF`](https://netlib.org/lapack/explore-html/df/dd6/group__gbtrf_ga725ad1dfee52c0dff571566d64e30ae8.html) which assumes that storage of size `2*lband+uband+1` is available for calculating the factorization.

In my case I was dealing with a tridiagonal Jacobian matrix. My function looks something like:
```
    def jacobian(self,t,u,banded=False):

        n = self.grid_size

        # The Jacobian is tridiagonal
        du = np.zeros(n)
        dc = np.zeros(n)
        dl = np.zeros(n)

        # ... initialize arrays

        if banded:
            ab = np.empty((3,n),order='F',dtype=u.dtype)
            for i in range(0,n):
                ab[0,i] = du[i]
                ab[1,i] = dc[i]
                ab[2,i] = dl[i]
            return ab
        else:
            jac = sp.sparse.dia_array(
                ((dl,dc,du),(-1,0,1)),
                shape=(n,n))
            return jac
```

When I use `solve_ivp` with the settings
```
    sol = solve_ivp(..., 
        method='LSODA',
        lband=1,uband=1,
        jac=lambda t,y: jacobian(t,y,banded=True))
```
I get a crash at runtime,
```
rv_cb_arr is NULL
Call-back cb_jac_in_lsoda__user__routines failed.
Fatal Python error: F2PySwapThreadLocalCallbackPtr: F2PySwapThreadLocalCallbackPtr: PyLong_AsVoidPtr failed
Python runtime state: initialized
ValueError: 0-th dimension must be fixed to 4 but got 3


Current thread 0x00007ff8541a1dc0 (most recent call first):
  File "/usr/local/Caskroom/miniconda/base/lib/python3.10/site-packages/scipy/integrate/_ode.py", line 1347 in run
  File "/usr/local/Caskroom/miniconda/base/lib/python3.10/site-packages/scipy/integrate/_ivp/lsoda.py", line 161 in _step_impl
  File "/usr/local/Caskroom/miniconda/base/lib/python3.10/site-packages/scipy/integrate/_ivp/base.py", line 197 in step
  File "/usr/local/Caskroom/miniconda/base/lib/python3.10/site-packages/scipy/integrate/_ivp/ivp.py", line 650 in solve_ivp
  File "/Users/ivan/tridiag.py", line 195 in <module>
...
```
When I add an extra row to the Jacobian matrix
```
    ab = np.empty((4,n), ...)
```
it appears to work. I made the `banded` branch print a message, and it appears in the output, so the routine is being called.

**Comment by MatthewFlamm** [comment]
Thanks for confirming this!  My memory from the added edit in https://github.com/scipy/scipy/issues/10793#issuecomment-530866735 is that larger systems with different bandedness is that it wasn't this simple.  But maybe it is?  There is some magic done for `odeint` which may be a clue.  I'm not convinced however that the `odeint` implementation is correct either.

**Comment by MatthewFlamm** [comment]
I'm reading your comment again, and this one I skipped over on the first read:

> It is consistent at least with the workspace requirements of the LU factorization routine [DGBTRF](https://netlib.org/lapack/explore-html/df/dd6/group__gbtrf_ga725ad1dfee52c0dff571566d64e30ae8.html) which assumes that storage of size 2*lband+uband+1 is available for calculating the factorization.

This is a great observation, and would explain why the banded-ness changes the storage requirement and expected structure of the jacobian passed in.  My reading of DGBTRF is that you set rows `KL+1: 2*KL+KU+1` and rows `1:KL` are 'blank', note Fortran is 1-based indexing.

I also have lots of problems with sparse arrays for Jacobians in scipy `solve_ivp`.  This is a separate issue.

**Comment by MatthewFlamm** [comment]
Some of the mystery is coming together here, but it raises more mystery: scipy does not use the vanilla LSODA and overrides the banded solver from this commit to be DGBTRF (as @ivan-pi pointed out):

https://github.com/scipy/scipy/commit/a678bf5eaa0011f7c23cd94f7c2079f8ba252e51

However, DGBTRF states that the empty data should come first while we have found that the empty data has to come last.  I have used my test case in the original description with the blank data first, and the # of function evaluations and # LU's jumps dramatically.  The result with the blank data last, matches the full dense Jacobian case in terms of function evals and LUs.  This is good evidence that this is how it should be supplied even if it does not match expectation from DGBTRF.  More work needs to be done in comparing different band structures.

`odeint` requires a completely packed matrix (doesn't matter if the blank data is first or last).  If you try to pass a [5, 5] matrix to `odeint` (from my original example in the PR description), it will throw an error:

```
In [37]: def band_jac(t, y):
    ...:     return [
    ...:      [0, 0, 0, 0, 0],
    ...:         [0,     0,                    0,  1e4*y[2], 0],
    ...:         [0,     0,             1e4*y[3], -1e4*y[2], 0],
    ...:         [1, -0.04, -1e4*y[3]-3e7*2*y[2],         0, 1],
    ...:         [0,  0.04,           3e7*2*y[2],         0, 0]
    ...:     ]
    ...:
In [38]: y_odeint_band = odeint(f, y0, t_eval, tfirst=True, Dfun=band_jac, ml=1, mu=2)
    ...: assert np.allclose(sol.y, y_odeint_band.T)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[38], line 1
----> 1 y_odeint_band = odeint(f, y0, t_eval, tfirst=True, Dfun=band_jac, ml=1, mu=2)
      2 assert np.allclose(sol.y, y_odeint_band.T)

File /workspaces/scipy/build-install/lib/python3.12/site-packages/scipy/integrate/_odepack_py.py:254, in odeint(func, y0, t, args, Dfun, col_deriv, full_output, ml, mu, rtol, atol, tcrit, h0, hmax, hmin, ixpr, mxstep, mxhnil, mxordn, mxords, printmessg, tfirst)
    251 y0 = copy(y0)
    253 with ODE_LOCK:
--> 254     output = _odepack.odeint(func, y0, t, args, Dfun, col_deriv, ml, mu,
    255                             full_output, rtol, atol, tcrit, h0, hmax, hmin,
    256                             ixpr, mxstep, mxhnil, mxordn, mxords,
    257                             int(bool(tfirst)))
    258 if output[-1] < 0:
    259     warning_msg = (f"{_msgs[output[-1]]} Run with full_output = 1 to "
    260                    f"get quantitative information.")

RuntimeError: Expected a banded Jacobian array with shape (4, 5)
```

When using the completely packed banded Jacobian in `odeint`, it doesn't exactly match the performance of `solve_ivp` using the banded Jacobian, but they are close.  This is a small test problem however, and we may see them diverge if a larger problem is considered.  This is another avenue to explore.

**Comment by MatthewFlamm** [comment]
@lucascolley  I'm not sure the title rename is correct here.  I originally found that `odeint` was working, but `solve_ivp` / `ode` implementations did not work. This was a result of me first using `odeint` and then transitioning to `solve_ivp` when it was introduced.

But we have found that it is _possible_ that all of them work, but the requirements for the packing structure for the banded matrix is opaque and hard to figure out.  I will change the title to be more precise to what I think is going on at this point in time.

**Comment by ivan-pi** [comment]
> However, DGBTRF states that the empty data should come first while we have found that the empty data has to come last. I have used my test case in the original description with the blank data first, and the # of function evaluations and # LU's jumps dramatically. The result with the blank data last, matches the full dense Jacobian case in terms of function evals and LUs. This is good evidence that this is how it should be supplied even if it does not match expectation from DGBTRF. More work needs to be done in comparing different band structures.

Yep, that matches my own testing. 👍🏻 I found this a bit suspicious, but I didn't dig deeper to find out if LSODA or the SciPy wrapper is shifting the rows into the position expected by the DGBTRF. Otherwise I don't see how this could work. 

I was sensitive to this issue, because in the past I noticed the difference in how `DGBMV` and `DGBTRF` need to be called (https://fortran-lang.discourse.group/t/help-with-blas-band-matrix-operations-solved/1150) - one has to pass the first element and leading dimension correctly, else you get a stair-case effect. An illustration of the leading dimension argument passing convention found in here (pg. 189 - 193): https://doku.lrz.de/dyn/Doku_Kurse/Fortran/basics/Fortran_3days.pdf

**Comment by ivan-pi** [comment]
On the other hand, I'm guessing the days of LSODA are limited. Discussion here: https://github.com/scipy/scipy/issues/18566

**Comment by MatthewFlamm** [comment]
I created a more comprehensive test in this gist below where I tested this with different size and # of bands on lower and upper side.  The solution of adding `kl`/`lband`/`mu` blank rows seems robust for `solve_ivp`, and thus also for `ode`, versions of LSODA,  The trends of # of function evaluations and # of Jacobian evaluations make sense to me.  My question is whether this should now be documented for LSODA in `solve_ivp` without being able to pin down why it is like this.  It is unexpected to me that it should be padded with empty data at the end.

https://gist.github.com/MatthewFlamm/825547630fc868ce952d5da55922dde0

**Comment by ivan-pi** [comment]
If I had to guess the error message,
> ValueError: 0-th dimension must be fixed to 4 but got 3

was coming from here:
https://github.com/scipy/scipy/blob/a32bc49b738db04e5823b676be40bd8dab617f8c/scipy/linalg/flapack_gen_banded.pyf.src#L43-L44

or here,

https://github.com/scipy/scipy/blob/a32bc49b738db04e5823b676be40bd8dab617f8c/scipy/linalg/flapack_gen_banded.pyf.src#L14-L18

The LSODA f2py header, doesn't seem capable of checking the consistency of `ldab` (in the banded solver routines) with `ml`/`mu` passed via `solve_ivp`.

In the C wrapper of the LSODA Jacobian User Function, a copy is always made in the banded case:

https://github.com/scipy/scipy/blob/a32bc49b738db04e5823b676be40bd8dab617f8c/scipy/integrate/_odepackmodule.c#L370

**Comment by MatthewFlamm** [comment]
This does looks like, at least on location, where the check happens for the size of the matrix utilized in the banded solver for `solve_ivp`

https://github.com/scipy/scipy/blob/a32bc49b738db04e5823b676be40bd8dab617f8c/scipy/linalg/flapack_gen_banded.pyf.src#L14-L18

I don't see where it reorders the matrix to switch the blank lines from bottom to top.  `dgbsv` in the code still says it starts at the `klth` row.

https://github.com/scipy/scipy/blob/a32bc49b738db04e5823b676be40bd8dab617f8c/scipy/linalg/flapack_gen_banded.pyf.src#L5-L10

**Comment by MatthewFlamm** [comment]
`solve_banded` in `scipy.linalg` uses the completely packed form.  It does the internal padding inside.    I wonder if this is involved in the weirdness here.  If this function is involved for LSODA in `solve_ivp` it would reorder the blank spaces.  It is also possible that the size of the matrix from the original jacobian is checked as part of the sequence leading to this behavior.

https://github.com/scipy/scipy/blob/d7979756b7a5b61e4320fd8a5a492238cdfdd06e/scipy/linalg/_basic.py#L534-L548

It is inefficient, and unexpected, that `solve_banded` will reorder matrices that are already supplied in the padded form.

**Comment by MatthewFlamm** [comment]
This was probably a red herring.  `solve_ivp` claims it uses the packing logic from `solve_banded`, but I cannot find anywhere that `solve_ivp` or the underlying `ode` version of LSODA uses `solve_banded.

https://github.com/scipy/scipy/blob/d7979756b7a5b61e4320fd8a5a492238cdfdd06e/scipy/integrate/_ivp/ivp.py#L348-L357

**Comment by ivan-pi** [comment]
That's a very good observation. 

If we borrow the DGBTRF image for a second `KL = 2, KU = 1`,
```txt
      *    *    *    +    +    +       *    *    *   u14  u25  u36
      *    *    +    +    +    +       *    *   u13  u24  u35  u46
      *   a12  a23  a34  a45  a56      *   u12  u23  u34  u45  u56
     a11  a22  a33  a44  a55  a66     u11  u22  u33  u44  u55  u66
     a21  a32  a43  a54  a65   *      m21  m32  m43  m54  m65   *
     a31  a42  a53  a64   *    *      m31  m42  m53  m64   *    *
```
the places marked by `+` are used for the factorization, and the `*` are ignored.
 
Let's now focus on what we observe in the SciPy LSODA wrapper. In Python we fill the values like this, 
```txt
      *   a12  a23  a34  a45  a56
     a11  a22  a33  a44  a55  a66
     a21  a32  a43  a54  a65   * 
     a31  a42  a53  a64   *    * 
      *    *    +    +    +    + 
      *    +    +    +    +    + 
```
Since Fortran arrays are stored column-major, this is actually equivalent to storing it like this
```txt
    wm(3)  *    *    +    +    +   +
           *    +    +    +    +   +
      *   a12  a23  a34  a45  a56       
     a11  a22  a33  a44  a55  a66
     a21  a32  a43  a54  a65   * 
     a31  a42  a53  a64   *    * 
```
In both cases, the leading dimension is `meband = 2*ml + mu + 1`. The documentation in `prja` state the matrix is stored at `wm(3)` so that `wm(3+ml)` would be just the star above `a11`. This could explain how things work "correctly." (Just not sure if it's done like this deliberately.)

**Comment by ivan-pi** [comment]
There is definitely something to be corrected, either in the `LSODA` docs, or in the ODEPACK extension module. But I think with the stated plans to replace LSODA (cc @ilayn), this is not worth pursuing.

**Comment by MatthewFlamm** [comment]
I think the ODEPACK extension module is for `odeint`, but this `lsoda.f` and `prja.f`, etc. are common to both.  This makes any fix inside `prja.f` for this issue problematic.  

I can think of some fixes, but I'm not sure any are ideal given comments about moving away from using the Fortran code.

1) Change the index for `wm` in the call to `dgbtrf` when there is a user supplied banded Jacobian, but this would likely break `odeint` in this case.

2) separate the Fortran LSODA.f for odeint and solve_ivp/ode and apply option 1.

3)  Borrow the wrapping of the user supplied Jacobian from `odeint` here.  

4) Rewrite `solve_ivp` to utilize `odeint` directly.  This would be difficult as `odeint` interface is much different than `solve_ivp` expects.

5) Accept this is how it works and document it.  Make it a feature.

6) Intercept the user supplied Jacobian and inject the necessary padding.

**Comment by dschmitz89** [comment]
So far most Fortran refactoring were literal translations. So I see the danger that we might just copy the bug over to the new (C?) translation. How difficult would it be to fix this? Do we need to work on the F77 base?

**Comment by MatthewFlamm** [comment]
> So far most Fortran refactorings were literal translations. So I see the danger that we might just copy the bug over to the new (C?) translation. How difficult would it be to fix this? Do we need to work on the F77 base?

Since this bug is particular to Fortran and Python memory layout and indexing, I think it isn't a likely bug to leep across in the translation into C.  There is also a quick test that can be written and the gist I linked to above can be simplified into a test for functionality.

There is a potential quick fix in the F77 code, see my comment above, by adjusting the indexing for this case.  However it would likely break `odeint`, which also uses this code, but with its own separate interface.

**Comment by ivan-pi** [comment]
> I think the ODEPACK extension module is for `odeint`, but this `lsoda.f` and `prja.f`, etc. are common to both. This makes any fix inside `prja.f` for this issue problematic.

The `LSODA` class (an extension of the `OdeSolver` class) used in `solve_ivp` is just a thin wrapper of `ode`, which wraps `_lsoda` (an f2py-generated extension module):

https://github.com/scipy/scipy/blob/9166b6fc4714fa55d4e829c8a76a278b8ecaf647/scipy/integrate/_ivp/lsoda.py#L2-L7

`odeint` is slightly different, because it is implemented as a wrapper to `LSODA` directly, without intermediary wrappers:

https://github.com/scipy/scipy/blob/9166b6fc4714fa55d4e829c8a76a278b8ecaf647/scipy/integrate/_odepackmodule.c#L718

Anyways, LSODA seems to work, it's just the inconsistency between the banded jacobian layout which differs between `LSODA`, `ode`, and `odeint`.

**Comment by ivan-pi** [comment]
> I can think of some fixes, but I'm not sure any are ideal given comments about moving away from using the Fortran code.

Number 5. would be the easiest. Instead of thinking for fixes, I would ask first what is the desired Python API (with padding or without padding). `solve_ivp` refers to the documentation of `solve_banded` which uses no padding:
```txt
     [*  * -1 -1 -1]
ab = [*  2  2  2  2]
     [5  4  3  2  1]
     [1  1  1  1  *]
```

If you look into the documentation of `solve_banded` it has to make a copy to accommodate the factorization (the `overwrite_ab` argument is meaningless...):

https://github.com/scipy/scipy/blob/6e246d0b54dd55dc69232a0caae6772228a7ac25/scipy/linalg/_basic.py#L632-L636

But since LSODA isn't using `solve_banded` this doesn't actually matter. And the other Python ODE solvers don't support the LAPACK banded format either. Only LSODA does.

I'm only 80 % certain of this, but the only way to avoid the making an internal copy of the Jacobian is if you add the padding on top and use `order=F`. Otherwise you won't have the right layout for `DGBTRF` and a copy will be required. (In principle you could also use C order, but then you need to allocate the banded array of size `n * (2*ml+mu+1)`, i.e. tall and slim.) But actually even this doesn't help avoid a copy, because the callback format is:
```python
def banded_jac(t,y,mu,ml) -> np.ndarray:
    ab = np.empty((mu+ml+1),len(y))    # allocate memory
    # ... fill ab ...
    return ab
    
sol = solve_ivp(..., jac = lambda t, y: banded_jac(t,y,ml,mu),lband=ml,uband=mu, ...)
```
whereas the `jac` callback within `prja` expects to receive the memory in the workspace memory:
```fortran
 400  ml = iwm(1)
      mu = iwm(2)
      ml3 = ml + 3
      mband = ml + mu + 1
      meband = mband + ml
      lenp = meband*n
      do 410 i = 1,lenp
 410    wm(i+2) = 0.0d0
      call srcma (rsav, isav, 1)
      call jac (neq, tn, y, ml, mu, wm(ml3), meband)    ! jacobian goes into wm
```
So there is no way of avoiding the banded Jacobian array copy, unless you change the SciPy callback to a mutating one:
```
def jac(t,y,dFdy) -> None:
    # ... fill the values of dFdy[i,j] in-place ...
```

Just my two cents.

**Comment by ilayn** [comment]
This whole `integrate` codebase is very confusing for me too. I don't quite get the intention with every part of the OOP parts (hence #21683 is still waiting). 

I did not have time to have a look thoroughly but it is possible that we have issues with the wrappers. The copies are done by f2py regardless of the input type. It converts all to the right layout and datatypes but it does not make any padding. So if the padding is needed we should add it in the wrappers. 

I'll try to come back to this when I have time but @j-bowhay also reported a gbsv segfault and I think it is in there somewhere as my first guess since the wrappers are quite dated by now.

**Comment by ilayn** [comment]
> Nifty. 😎

Possibly, but as you encountered, requires days of investigation for an epsilon*`ml` gain.

**Comment by MatthewFlamm** [comment]
> @MatthewFlamm, so in odeint no extra rows are needed?

You can see from the original issue description code that it works without padding, as well as the gist code from https://github.com/scipy/scipy/issues/10793#issuecomment-2573975225.  In https://github.com/scipy/scipy/issues/10793#issuecomment-2573484437, `odeint` will reject padding (either on top or bottom).

Here is where I think `odeint` does the memory copy to Fortran.  `jac_type == 4` is the banded user supplied Jacobian case we are discussing.  I can't follow everything, but it looks like it puts the user array at the beginning of the Fortran array in `copy_array_to_fortran` which means it is padded at the end as expected.

https://github.com/scipy/scipy/blob/9166b6fc4714fa55d4e829c8a76a278b8ecaf647/scipy/integrate/_odepackmodule.c#L363-L373

**Comment by MatthewFlamm** [comment]
`vode` also claims to accept a banded Jacobian.  I haven't tried it yet, but will report back when I do.  If it accepts the fully packed version, it would give extra weight to enforcing it for lsoda.  In addition, I expect this was the original intention by the documentation).

If we need to memory copy anyways, what's another one? ;).  One (possibly temporary) solution is to intercept the user supplied Jacobian with fully packed structure and pad it with the zeros at the beginning.  This adds yet another copy, but would fix the bug while a better fix can be made.

**Comment by ivan-pi** [comment]
I think VODE should work because I noticed it has a shim function which pads the Jacobian:


https://github.com/scipy/scipy/blob/9166b6fc4714fa55d4e829c8a76a278b8ecaf647/scipy/integrate/_ode.py#L1006-L1010

It seems like this wrapper is missing for LSODA.

**Comment by MatthewFlamm** [comment]
There is already precedent for this double memory copy idea then. Everything has been done before :)

**Comment by ivan-pi** [comment]
> Here is where I think `odeint` does the memory copy to Fortran. `jac_type == 4` is the banded user supplied Jacobian case we are discussing. I can't follow everything, but it looks like it puts the user array at the beginning of the Fortran array in `copy_array_to_fortran` which means it is padded at the end (added: in the Fortran array, but not the user supplied array) as expected.

I agree here. It copies the data into the top rows, but the leading dimension of the `result_array` is `nrowpd` which accomodates the extra scratch space for the factorization. 

We are getting somewhere. 👍🏻

**Comment by ivan-pi** [comment]
> I think VODE should work because I noticed it has a shim function which pads the Jacobian (just like you say in your post a minute ago):
> 
> https://github.com/scipy/scipy/blob/9166b6fc4714fa55d4e829c8a76a278b8ecaf647/scipy/integrate/_ode.py#L1006-L1010
> 
> It seems like this wrapper is missing for `class lsoda(ode)`.

So this comment implies it is a requirement of the f2py wrappers, either the `_lsoda -> jac` one or the LAPACK `DGBTRF` one:

https://github.com/scipy/scipy/blob/a32bc49b738db04e5823b676be40bd8dab617f8c/scipy/linalg/flapack_gen_banded.pyf.src#L43-L44

which is why we get the error message as I suspected. By padding the layout (and hence the checks) meet the LSODA requirements.

@MatthewFlamm would you like to write the (temporary?) fix? I think all you have to do is add the banded Jacobian padding (in the spirit of the [VODE one]( https://github.com/scipy/scipy/blob/9166b6fc4714fa55d4e829c8a76a278b8ecaf647/scipy/integrate/_ode.py#L1010)) in front of this lines:

https://github.com/scipy/scipy/blob/9166b6fc4714fa55d4e829c8a76a278b8ecaf647/scipy/integrate/_ode.py#L1354-L1355

**Comment by MatthewFlamm** [comment]
> @MatthewFlamm would you like to write the (temporary?) fix? I think all you have to do is add the banded Jacobian padding (in the spirit of the VODE one) in front of these lines:

I will do it. Thanks for the enlightening discussion on this issue that has plagued me for many years!

**Comment by ivan-pi** [comment]
I will contribute my example as a test. :)

**Comment by ivan-pi** [comment]
I'm a bit surprised the lsoda error isn't caught in the recently added tests by @WarrenWeckesser: https://github.com/scipy/scipy/commit/7a0c1676bd7b1afef6dec0e2fe71e9e50e8956ec,

specifically the lines here:

https://github.com/scipy/scipy/blob/7a0c1676bd7b1afef6dec0e2fe71e9e50e8956ec/scipy/integrate/tests/test_banded_ode_solvers.py#L178-L185

Is there a place I can check these in the CI logs or do I need to run it locally? (Sorry for the naive question - my first time digging around here.)

**Comment by MatthewFlamm** [comment]
There is also a test here

https://github.com/scipy/scipy/blob/cbe4535abf233d935a991ac4b76c22e9e3e81d17/scipy/integrate/tests/test_integrate.py#L470-L474

that creates a fully packed banded Jacobian

https://github.com/scipy/scipy/blob/cbe4535abf233d935a991ac4b76c22e9e3e81d17/scipy/integrate/tests/test_integrate.py#L497-L501

The problem is that these tests are fatally flawed.  The test in `test_banded_ode_solvers` is clearly nonstiff.  The test in `test_integrate` is not stiff enough to get `lsoda` to do an implicit step, and thus there are no `jac` calls.

> Is there a place I can check these in the CI logs or do I need to run it locally? (Sorry for the naive question - my first time digging around here.)

This won't be in the CI logs because everything passes.  I had to go into the tests themselves and poke at things.

**Comment by MatthewFlamm** [comment]
The test in `test_banded_solvers` is complex with an analytical solution check.  My thought is to replace that test entirely with a stiffer problem, e.g. one of the ones we have supplied, and check the solution against the full Jacobian evaluation.

**Comment by ivan-pi** [comment]
I see. My problem is mildly stiff (diffusion problem), and the Jacobian is invoked about ~10 times on my given time interval. A bit surprised that there are not more ODE tests (there are many in the University of Bari test set - https://archimede.uniba.it/~testset/).

**Comment by ilayn** [comment]
I just finished LSODA translation in #23914 and the very first example at the top of this issue gives 

```
RuntimeError                              Traceback (most recent call last)
Cell In[3], line 45
     42 y_odeint = odeint(f, y0, t_eval, tfirst=True, Dfun=jac)
     43 assert np.allclose(sol.y, y_odeint.T)
---> 45 y_odeint_band = odeint(f, y0, t_eval, tfirst=True, Dfun=band_jac, ml=1, mu=2)
     46 assert np.allclose(sol.y, y_odeint_band.T)
     48 sol = solve_ivp(f, tspan, y0, lband=1, uband=2, jac=band_jac, t_eval=t_eval, method='LSODA', rtol=rtol, atol=atol)

File ~/Documents/GitHub/scipy/build-install/usr/lib/python3.13/site-packages/scipy/integrate/_odepack_py.py:252, in odeint(func, y0, t, args, Dfun, col_deriv, full_output, ml, mu, rtol, atol, tcrit, h0, hmax, hmin, ixpr, mxstep, mxhnil, mxordn, mxords, printmessg, tfirst)
    248 t = copy(t)
    249 y0 = copy(y0)
--> 252 output = _odepack.odeint(func, y0, t, args, Dfun, col_deriv, ml, mu,
    253                         full_output, rtol, atol, tcrit, h0, hmax, hmin,
    254                         ixpr, mxstep, mxhnil, mxordn, mxords,
    255                         int(bool(tfirst)))
    256 if output[-1] < 0:
    257     warning_msg = (f"{_msgs[output[-1]]} Run with full_output = 1 to "
    258                    f"get quantitative information.")

RuntimeError: Expected a Jacobian array with shape (5, 5), but got (4, 5)
```

So apparently the bug is carried over but I think this is a requirement from the LSODA code and not from the wrappers. I tried many other stiff problems (dense and banded) and the results are in agreement with the Fortran cases. However the code is one of the worst I had encountered even for Fortran 77 standards. Hence, I'm not sure if it is worth anybody's time to inspect the issue here. But if you know a concrete fix we can include it.

**Comment by WarrenWeckesser** [comment]
@ilayn, I'm confused.  This issue is about `solve_ivp`.  Your example shows an error being raised by `odeint`, so isn't that unrelated to this issue?

**Comment by ilayn** [comment]
I was confused for a while too. They come from the same codebase. Previously, for some weird reason, we wrapped LSODA both as `odeint` and separately as LSODA (you can select LSODA from `solve_ivp` interface with the keyword). Not sure why. There are some layers of historical crust in integrate. We should wrap SUNDIALS and get rid of the old versions of these since it is well-maintained in C code.

**Comment by WarrenWeckesser** [comment]
Sure, there is messy history, but I still don't understand why you have commented in this issue.  You got the error `RuntimeError: Expected a Jacobian array with shape (5, 5), but got (4, 5)`.  If that error is being generated by `odeint` with your new translation, then there is a problem with your translation.  With state space size of 5, `ml=1`, and `mu=2`, the user-provided function `Dfun` must return an array with shape (4, 5).

**Comment by ilayn** [comment]
Am I misunderstanding that this is an issue of LSODA forcing users to submit full nxn instead of a banded storage jacobian ?

**Comment by WarrenWeckesser** [comment]
I suspect something is broken in your new wrapper that makes it expect a full Jacobian, even though `Dfun`, `ml` and `mu` have been provided.

Further discussion should continue over in the PR.  I probably won't be able to do a deep dive into it any time soon, but after I get a bunch of other tasks off my SciPy & NumPy "TO DO" lists I can at least build it and run it through its paces.

**Comment by MatthewFlamm** [comment]
It seems to me that the rewrite has created the same problem that `solve_ivp` and `ode` used to have before #22283, but now for `odeint`.  Note that 5 rows is `2*ml+mu+1` for the example.  It is a red herring that it matches the full Jacobian size.

The gist in https://github.com/scipy/scipy/issues/10793#issuecomment-2573975225 has some in depth testing of different band structures including the fully packed form (ml+mu+1) and the padded form (2*ml+mu+1).  It could be useful for testing whether my hypothesis is correct.

If so, then the new c lsoda code needs to take in a fully packed user function from Python, and then pass the padded form to the c code.  This is how the `odeint` interface used to work.  If this is implemented, then `solve_ivp` might break.  The padding in #22283 could be removed to fix `solve_ivp`.

**Comment by ilayn** [comment]
Yes unfortunately there are too many entry points to the same code and each has different quirks, due to historical build-up. I still don't know if I handled all of them. I'll add your tests in your gist to the interface tests and make them work. Thanks.

**Comment by MatthewFlamm** [comment]
The multiple entry points is indeed confusing.  The proposed PR is nice that it reduces one level of complexity.

To be clear on the use of the gist code, it was used to test how the code worked prior to #22283.  It will fail after the fix was put in, but it should pass if the fully packed form is used for the Jacobian in `solve_ivp` after #22283.

**Comment by MatthewFlamm** [comment]
I'm assuming the error is coming from this code:

https://github.com/ilayn/scipy/blob/b8703fabd2ec76af764d365fcf8edbdea3bad006/scipy/integrate/_odepackmodule.c#L362-L373

I'm reading that it should have the word "banded" in the error message if correctly identified as banded, but it doesn't show up in your error message.
